Use Proguard rules for Thunderbird release builds

Currently the release build of `:app-thunderbird` doesn't include the `proguard-rules.pro` file. This leads to R8 stripping too much code and the app crashing when setting up an IMAP account.

Steps to reproduce:
1. `./gradlew :app-thunderbird:assembleRelease` (make sure a keystore is available so a signed APK is created)
2. `adb install app-thunderbird/build/outputs/apk/release/app-thunderbird-release.apk`
3. Set up an IMAP account

Stack trace:
```text
java.lang.AssertionError: Missing field in com.fsck.k9.mail.Flag
    at com.squareup.moshi.StandardJsonAdapters$EnumJsonAdapter.<init>(SourceFile:92)
    at com.squareup.moshi.MapJsonAdapter$1.create(SourceFile:514)
    at com.squareup.moshi.Moshi.adapter(SourceFile:29)
    at com.squareup.moshi.MapJsonAdapter$1.create(SourceFile:1069)
    at com.squareup.moshi.Moshi.adapter(SourceFile:29)
    at com.fsck.k9.controller.PendingCommandSerializer.<init>(SourceFile:102)
    at com.fsck.k9.controller.PendingCommandSerializer.<clinit>(SourceFile:3)
    at com.fsck.k9.mailstore.LocalStore.<init>(SourceFile:6)
    at com.fsck.k9.mailstore.LocalStoreProvider.getInstance(SourceFile:48)
    at com.fsck.k9.mailstore.SpecialLocalFoldersCreator.createSpecialLocalFolders(SourceFile:13)
    at com.fsck.k9.account.AccountCreator$createAccount$2.invokeSuspend(SourceFile:485)
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(SourceFile:9)
    at kotlinx.coroutines.DispatchedTask.run(SourceFile:113)
    at androidx.work.Worker$2.run(SourceFile:13)
    at kotlinx.coroutines.scheduling.TaskImpl.run(SourceFile:3)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(SourceFile:96)
    Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@bfda7dd, Dispatchers.Main.immediate]
Caused by: java.lang.NoSuchFieldException: RECENT
    at java.lang.Class.getField(Class.java:2337)
    at com.squareup.moshi.StandardJsonAdapters$EnumJsonAdapter.<init>(SourceFile:33)
    ... 15 more
```

--- a/.gitignore
+++ b/.gitignore
@@ -12,6 +12,7 @@ gen/
 out/
 
 # Generated files
+*.aab
 *.apk
 *.ap_
 *.class

--- a/app-k9mail/build.gradle.kts
+++ b/app-k9mail/build.gradle.kts
@@ -106,7 +106,7 @@ android {
     }
 
     signingConfigs {
-        createSigningConfig(project, SigningType.K9_RELEASE)
+        createSigningConfig(project, SigningType.K9_RELEASE, isUpload = false)
     }
 
     buildTypes {

--- a/app-thunderbird/build.gradle.kts
+++ b/app-thunderbird/build.gradle.kts
@@ -128,8 +128,14 @@ android {
 
         release {
             signingConfig = signingConfigs.findByName("release")
+
             isMinifyEnabled = true
             isShrinkResources = true
+
+            proguardFiles(
+                getDefaultProguardFile("proguard-android.txt"),
+                "proguard-rules.pro",
+            )
         }
 
         create("daily") {

--- a/app-thunderbird/build.gradle.kts
+++ b/app-thunderbird/build.gradle.kts
@@ -106,18 +106,9 @@ android {
     }
 
     signingConfigs {
-        if (project.hasProperty("thunderbird.keyAlias") &&
-            project.hasProperty("thunderbird.keyPassword") &&
-            project.hasProperty("thunderbird.storeFile") &&
-            project.hasProperty("thunderbird.storePassword")
-        ) {
-            create("release") {
-                keyAlias = project.property("thunderbird.keyAlias") as String
-                keyPassword = project.property("thunderbird.keyPassword") as String
-                storeFile = file(project.property("thunderbird.storeFile") as String)
-                storePassword = project.property("thunderbird.storePassword") as String
-            }
-        }
+        createSigningConfig(project, SigningType.TB_RELEASE)
+        createSigningConfig(project, SigningType.TB_BETA)
+        createSigningConfig(project, SigningType.TB_DAILY)
     }
 
     buildTypes {
@@ -129,7 +120,7 @@ android {
         }
 
         release {
-            signingConfig = signingConfigs.findByName("release")
+            signingConfig = signingConfigs.getByType(SigningType.TB_RELEASE)
 
             isMinifyEnabled = true
             isShrinkResources = true
@@ -140,20 +131,24 @@ android {
             )
         }
 
-        create("daily") {
+        create("beta") {
             initWith(getByName("release"))
 
-            applicationIdSuffix = ".daily"
-            versionNameSuffix = "a1"
+            signingConfig = signingConfigs.getByType(SigningType.TB_BETA)
+
+            applicationIdSuffix = ".beta"
+            versionNameSuffix = "b1"
 
             matchingFallbacks += listOf("release")
         }
 
-        create("beta") {
+        create("daily") {
             initWith(getByName("release"))
 
-            applicationIdSuffix = ".beta"
-            versionNameSuffix = "b1"
+            signingConfig = signingConfigs.getByType(SigningType.TB_DAILY)
+
+            applicationIdSuffix = ".daily"
+            versionNameSuffix = "a1"
 
             matchingFallbacks += listOf("release")
         }

--- a/app-thunderbird/build.gradle.kts
+++ b/app-thunderbird/build.gradle.kts
@@ -43,7 +43,7 @@ android {
         testApplicationId = "net.thunderbird.placeholder.tests"
 
         versionCode = 1
-        versionName = "0.1-SNAPSHOT"
+        versionName = "0.1"
 
         // Keep in sync with the resource string array "supported_languages"
         resourceConfigurations.addAll(
@@ -123,6 +123,8 @@ android {
     buildTypes {
         debug {
             applicationIdSuffix = ".debug"
+            versionNameSuffix = "-SNAPSHOT"
+
             isMinifyEnabled = false
         }
 
@@ -140,13 +142,19 @@ android {
 
         create("daily") {
             initWith(getByName("release"))
+
             applicationIdSuffix = ".daily"
+            versionNameSuffix = "a1"
+
             matchingFallbacks += listOf("release")
         }
 
         create("beta") {
             initWith(getByName("release"))
+
             applicationIdSuffix = ".beta"
+            versionNameSuffix = "b1"
+
             matchingFallbacks += listOf("release")
         }
     }

--- a/app-thunderbird/build.gradle.kts
+++ b/app-thunderbird/build.gradle.kts
@@ -39,8 +39,8 @@ android {
     namespace = "net.thunderbird.android"
 
     defaultConfig {
-        applicationId = "net.thunderbird.placeholder"
-        testApplicationId = "net.thunderbird.placeholder.tests"
+        applicationId = "net.thunderbird.android"
+        testApplicationId = "net.thunderbird.android.tests"
 
         versionCode = 1
         versionName = "0.1"

--- a/app-thunderbird/build.gradle.kts
+++ b/app-thunderbird/build.gradle.kts
@@ -42,7 +42,7 @@ android {
         applicationId = "net.thunderbird.android"
         testApplicationId = "net.thunderbird.android.tests"
 
-        versionCode = 1
+        versionCode = 2
         versionName = "0.1"
 
         // Keep in sync with the resource string array "supported_languages"
@@ -117,6 +117,8 @@ android {
             versionNameSuffix = "-SNAPSHOT"
 
             isMinifyEnabled = false
+            isShrinkResources = false
+            isDebuggable = true
         }
 
         release {
@@ -124,6 +126,7 @@ android {
 
             isMinifyEnabled = true
             isShrinkResources = true
+            isDebuggable = false
 
             proguardFiles(
                 getDefaultProguardFile("proguard-android.txt"),
@@ -132,25 +135,39 @@ android {
         }
 
         create("beta") {
-            initWith(getByName("release"))
-
             signingConfig = signingConfigs.getByType(SigningType.TB_BETA)
 
             applicationIdSuffix = ".beta"
             versionNameSuffix = "b1"
 
+            isMinifyEnabled = true
+            isShrinkResources = true
+            isDebuggable = false
+
             matchingFallbacks += listOf("release")
+
+            proguardFiles(
+                getDefaultProguardFile("proguard-android.txt"),
+                "proguard-rules.pro",
+            )
         }
 
         create("daily") {
-            initWith(getByName("release"))
-
             signingConfig = signingConfigs.getByType(SigningType.TB_DAILY)
 
             applicationIdSuffix = ".daily"
             versionNameSuffix = "a1"
 
+            isMinifyEnabled = true
+            isShrinkResources = true
+            isDebuggable = false
+
             matchingFallbacks += listOf("release")
+
+            proguardFiles(
+                getDefaultProguardFile("proguard-android.txt"),
+                "proguard-rules.pro",
+            )
         }
     }
 

--- a/build-plugin/src/main/kotlin/SigningExtensions.kt
+++ b/build-plugin/src/main/kotlin/SigningExtensions.kt
@@ -6,6 +6,7 @@ import org.gradle.api.Project
 
 private const val SIGNING_FOLDER = ".signing"
 private const val SIGNING_FILE_ENDING = ".signing.properties"
+private const val UPLOAD_FILE_ENDING = ".upload.properties"
 
 private const val PROPERTY_STORE_FILE = "storeFile"
 private const val PROPERTY_STORE_PASSWORD = "storePassword"
@@ -16,7 +17,7 @@ private const val PROPERTY_KEY_PASSWORD = "keyPassword"
  * Creates an [ApkSigningConfig] for the given signing type.
  *
  * The signing properties are read from a file in the `.signing` folder in the project root directory.
- * File names are expected to be in the format `$app.$type.signing.properties`.
+ * File names are expected to be in the format `$app.$type.signing.properties` or `$app.$type.upload.properties`.
  *
  * The file should contain the following properties:
  * - `$app.$type.storeFile`
@@ -26,9 +27,14 @@ private const val PROPERTY_KEY_PASSWORD = "keyPassword"
  *
  * @param project the project to create the signing config for
  * @param signingType the signing type to create the signing config for
+ * @param isUpload whether the upload or signing config is used
  */
-fun NamedDomainObjectContainer<out ApkSigningConfig>.createSigningConfig(project: Project, signingType: SigningType) {
-    val properties = project.readSigningProperties(signingType)
+fun NamedDomainObjectContainer<out ApkSigningConfig>.createSigningConfig(
+    project: Project,
+    signingType: SigningType,
+    isUpload: Boolean = true,
+) {
+    val properties = project.readSigningProperties(signingType, isUpload)
 
     if (properties.hasSigningConfig(signingType)) {
         create(signingType.type) {
@@ -37,6 +43,8 @@ fun NamedDomainObjectContainer<out ApkSigningConfig>.createSigningConfig(project
             keyAlias = properties.getSigningProperty(signingType, PROPERTY_KEY_ALIAS)
             keyPassword = properties.getSigningProperty(signingType, PROPERTY_KEY_PASSWORD)
         }
+    } else {
+        println("Signing config not created for ${signingType.type}")
     }
 }
 
@@ -49,8 +57,13 @@ fun NamedDomainObjectContainer<out ApkSigningConfig>.getByType(signingType: Sign
     return findByName(signingType.type)
 }
 
-private fun Project.readSigningProperties(signingType: SigningType) = Properties().apply {
-    val signingPropertiesFile = rootProject.file("$SIGNING_FOLDER/${signingType.id}$SIGNING_FILE_ENDING")
+private fun Project.readSigningProperties(signingType: SigningType, isUpload: Boolean) = Properties().apply {
+    val signingPropertiesFile = if (isUpload) {
+        rootProject.file("$SIGNING_FOLDER/${signingType.id}$UPLOAD_FILE_ENDING")
+    } else {
+        rootProject.file("$SIGNING_FOLDER/${signingType.id}$SIGNING_FILE_ENDING")
+    }
+
     if (signingPropertiesFile.exists()) {
         FileInputStream(signingPropertiesFile).use { inputStream ->
             load(inputStream)

--- a/build-plugin/src/main/kotlin/SigningType.kt
+++ b/build-plugin/src/main/kotlin/SigningType.kt
@@ -4,4 +4,7 @@ enum class SigningType(
     val id: String = "$app.$type",
 ) {
     K9_RELEASE(app = "k9", type = "release"),
+    TB_RELEASE(app = "tb", type = "release"),
+    TB_BETA(app = "tb", type = "beta"),
+    TB_DAILY(app = "tb", type = "daily"),
 }

